state/machine.go

```diff
@@ -1612,6 +1612,30 @@ func (m *Machine) ProviderAddresses() (addresses corenetwork.SpaceAddresses) {
 	return
 }
 
+// ProviderAddressesBySpaceName groups the machine addresses by space name and
+// returns the result as a map where the space name is used a the key.
+func (m *Machine) ProviderAddressesBySpaceName() (map[string][]corenetwork.SpaceAddress, error) {
+	addresses, err := m.AllAddresses()
+	if err != nil {
+		return nil, err
+	}
+
+	res := make(map[string][]corenetwork.SpaceAddress)
+	for _, address := range addresses {
+		// TODO(achilleasa): currently, addresses do not come with
+		// resolved space IDs (except MAAS). For the time being, we
+		// need to pull out the space ID information from the
+		// associated subnet.
+		subnet, err := address.Subnet()
+		if err != nil {
+			return nil, err
+		}
+		spaceName := subnet.SpaceName()
```

Comment:
When endpoint bindings are stored with space IDs, this will change, but I am surprised it works now.

Adding a space with subnets only populates the subnet's space ID, so I expect the map to only have a single key for "". I think we need a space lookup with a note that it will be removed.

As an aside, we'll need to audit for `subnet.SpaceName()` and see if we can deprecate it properly.

Reply:
This was copied from `machine.AllSpaces()` which I used to validate that all unit machines have an address in the requested space before changing the bindings (see linked PR from description). This allows us to query space names even though the `AllAddresses`(?) calls do not yet provide a spaceID for non-MAAS substrates.

Reply:
It looks like this works on MAAS because the incoming `SubnetInfo` from the provider populates subnet space name. As we discussed, we can land this and it will be picked up as part of @hmlanigan's in-flight patch.